# MNIST Classification using CNN

Step 1 : Collect Data

In [25]:
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets('MNIST_data', one_hot=True)

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


Step 1a : Get Training and Test Data

In [0]:
trainX = mnist.train.images
trainY = mnist.train.labels

In [0]:
testX = mnist.test.images
testY = mnist.test.labels

Set the Seed

In [0]:
import numpy as np

In [0]:
np.random.seed(42)

In [30]:
trainX.shape

(55000, 784)

# Build the Graph

import Keras Models and Layers

In [0]:
from tensorflow.python.keras.models import Sequential

In [0]:
from tensorflow.python.keras.layers import Dense, Dropout, Conv2D, Flatten, Reshape, MaxPool2D

Start a Sequential model

In [0]:
model = Sequential()

Reshape Data

In [0]:
model.add(Reshape((28,28,1),input_shape=(784,)))

Add first Convolutional layer : 32 filters 3x3 with stride as 1x1

In [0]:
model.add(Conv2D(32,kernel_size=(5,5),activation='relu',strides=(1,1)))
#26x26x32

Add 2nd Convolutional layer : 64 filters 3x3 with stride as 1x1

In [0]:
model.add(Conv2D(64,kernel_size=(5,5),strides=(1,1),activation='relu'))
#24x24x64

Add MaxPooling layer

In [0]:
model.add(MaxPool2D(pool_size=(2,2)))
#6x6x64

Add Dropout

In [0]:
model.add(Dropout(0.25))

Flatten the Output

In [0]:
model.add(Flatten())

Add Fully Connected Layer - 128 neurons

In [0]:
model.add(Dense(128,activation='relu'))

Another Dropout

In [0]:
model.add(Dropout(0.5))
#Suppose you have 128 units in the layer and set the keep_prob to 0.5, Then the activation of 64 randomly chosen units out of 128 will be set to 0, and the remaining 64 will be scaled by a factor of 50. I think a more precise description is that you only keep the activation of 50 percent of the nodes.

Add Softmax Layer

In [0]:
model.add(Dense(10,activation='softmax'))

Compile model    

In [0]:
model.compile(optimizer='adam',loss='categorical_crossentropy',metrics=['accuracy'])

Train

In [43]:
model.fit(trainX,trainY,epochs=20,batch_size=500,
          validation_data=(testX,testY))

Train on 55000 samples, validate on 10000 samples
Epoch 1/20
55000/55000 [==============================] - 2s 36us/sample - loss: 0.5249 - acc: 0.8412 - val_loss: 0.1470 - val_acc: 0.9566
Epoch 2/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.1814 - acc: 0.9468 - val_loss: 0.0799 - val_acc: 0.9767
Epoch 3/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.1253 - acc: 0.9637 - val_loss: 0.0596 - val_acc: 0.9820
Epoch 4/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.1015 - acc: 0.9704 - val_loss: 0.0502 - val_acc: 0.9839
Epoch 5/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0843 - acc: 0.9755 - val_loss: 0.0431 - val_acc: 0.9863
Epoch 6/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0740 - acc: 0.9780 - val_loss: 0.0400 - val_acc: 0.9857
Epoch 7/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0678 - acc: 0.9795 - val_loss: 0.

3 x 3

rain on 55000 samples, validate on 10000 samples
Epoch 1/20
55000/55000 [==============================] - 8s 148us/sample - loss: 0.4360 - acc: 0.8666 - val_loss: 0.0901 - val_acc: 0.9728
Epoch 2/20
55000/55000 [==============================] - 5s 89us/sample - loss: 0.1204 - acc: 0.9642 - val_loss: 0.0483 - val_acc: 0.9842
Epoch 3/20
55000/55000 [==============================] - 5s 89us/sample - loss: 0.0815 - acc: 0.9760 - val_loss: 0.0398 - val_acc: 0.9866
Epoch 4/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0683 - acc: 0.9789 - val_loss: 0.0330 - val_acc: 0.9890
Epoch 5/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0562 - acc: 0.9827 - val_loss: 0.0315 - val_acc: 0.9890
Epoch 6/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0481 - acc: 0.9856 - val_loss: 0.0310 - val_acc: 0.9884
Epoch 7/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0425 - acc: 0.9868 - val_loss: 0.0296 - val_acc: 0.9906
Epoch 8/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0366 - acc: 0.9882 - val_loss: 0.0309 - val_acc: 0.9894
Epoch 9/20
55000/55000 [==============================] - 5s 89us/sample - loss: 0.0355 - acc: 0.9889 - val_loss: 0.0268 - val_acc: 0.9911
Epoch 10/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0332 - acc: 0.9892 - val_loss: 0.0303 - val_acc: 0.9899
Epoch 11/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0289 - acc: 0.9902 - val_loss: 0.0261 - val_acc: 0.9917
Epoch 12/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0276 - acc: 0.9909 - val_loss: 0.0259 - val_acc: 0.9920
Epoch 13/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0257 - acc: 0.9918 - val_loss: 0.0270 - val_acc: 0.9911
Epoch 14/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0239 - acc: 0.9922 - val_loss: 0.0276 - val_acc: 0.9915
Epoch 15/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0218 - acc: 0.9927 - val_loss: 0.0288 - val_acc: 0.9908
Epoch 16/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0191 - acc: 0.9936 - val_loss: 0.0287 - val_acc: 0.9919
Epoch 17/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0194 - acc: 0.9938 - val_loss: 0.0282 - val_acc: 0.9915
Epoch 18/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0185 - acc: 0.9940 - val_loss: 0.0284 - val_acc: 0.9919
Epoch 19/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0176 - acc: 0.9940 - val_loss: 0.0305 - val_acc: 0.9912
Epoch 20/20
55000/55000 [==============================] - 5s 88us/sample - loss: 0.0178 - acc: 0.9939 - val_loss: 0.0263 - val_acc: 0.9926
<tensorflow.python.keras.callbacks.History at 0x7fea279f13c8>

5 x 5
without dropout
Train on 55000 samples, validate on 10000 samples
Epoch 1/20
55000/55000 [==============================] - 2s 36us/sample - loss: 0.5249 - acc: 0.8412 - val_loss: 0.1470 - val_acc: 0.9566
Epoch 2/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.1814 - acc: 0.9468 - val_loss: 0.0799 - val_acc: 0.9767
Epoch 3/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.1253 - acc: 0.9637 - val_loss: 0.0596 - val_acc: 0.9820
Epoch 4/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.1015 - acc: 0.9704 - val_loss: 0.0502 - val_acc: 0.9839
Epoch 5/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0843 - acc: 0.9755 - val_loss: 0.0431 - val_acc: 0.9863
Epoch 6/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0740 - acc: 0.9780 - val_loss: 0.0400 - val_acc: 0.9857
Epoch 7/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0678 - acc: 0.9795 - val_loss: 0.0365 - val_acc: 0.9875
Epoch 8/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0608 - acc: 0.9817 - val_loss: 0.0358 - val_acc: 0.9871
Epoch 9/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0566 - acc: 0.9827 - val_loss: 0.0347 - val_acc: 0.9890
Epoch 10/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0538 - acc: 0.9834 - val_loss: 0.0327 - val_acc: 0.9890
Epoch 11/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0490 - acc: 0.9847 - val_loss: 0.0320 - val_acc: 0.9890
Epoch 12/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0464 - acc: 0.9862 - val_loss: 0.0293 - val_acc: 0.9902
Epoch 13/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0429 - acc: 0.9869 - val_loss: 0.0301 - val_acc: 0.9891
Epoch 14/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0417 - acc: 0.9871 - val_loss: 0.0296 - val_acc: 0.9900
Epoch 15/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0412 - acc: 0.9869 - val_loss: 0.0277 - val_acc: 0.9910
Epoch 16/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0381 - acc: 0.9880 - val_loss: 0.0296 - val_acc: 0.9899
Epoch 17/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0359 - acc: 0.9888 - val_loss: 0.0302 - val_acc: 0.9898
Epoch 18/20
55000/55000 [==============================] - 2s 29us/sample - loss: 0.0352 - acc: 0.9885 - val_loss: 0.0281 - val_acc: 0.9899
Epoch 19/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0317 - acc: 0.9899 - val_loss: 0.0275 - val_acc: 0.9910
Epoch 20/20
55000/55000 [==============================] - 2s 30us/sample - loss: 0.0307 - acc: 0.9904 - val_loss: 0.0269 - val_acc: 0.9913
<tensorflow.python.keras.callbacks.History at 0x7fea2030a320>


Saving Model as HDF5 file

In [0]:
model.save('keras_cnn.h5')

Show Model

In [45]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
reshape_1 (Reshape)          (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 24, 24, 32)        832       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 12, 12, 32)        0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 12, 12, 32)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 4608)              0         
_________________________________________________________________
dense_2 (Dense)              (None, 128)               589952    
_________________________________________________________________
dropout_3 (Dropout)          (None, 128)              